### PR TITLE
feat(python): broadcast scalars for optimizers

### PR DESCRIPTION
**Pull request by Tengxu-Sun**
initial implement

### CODE REVIEW

File: bagua/torch_api/distributed.py [hunk 1]
```diff
@@ -93,23 +98,100 @@ def produces_sparse_gradient(module):
...
+
+        # Initialize newly created optimizers.
+        if len(optimizer_state_dict["state"]) == 0:
+            for group in optimizer.param_groups:
+                for p in group["params"]:
+                    if p.requires_grad and id(p) not in optimizer_state_dict["state"]:
+                        p.grad = p.data.new(p.size()).zero_()
+                        if isinstance(optimizer, torch.optim.SparseAdam):
+                            p.grad = p.grad.to_sparse()
+            optimizer.step()
```
Comment (wangraying) [inline bagua/torch_api/distributed.py:117]: We must call `optimizer.step` to get the state dict?

File: bagua/torch_api/distributed.py [hunk 2]
```diff
@@ -93,23 +98,100 @@ def produces_sparse_gradient(module):
...
+        for key, param in params:
+            broadcast(param, src=0)
+        scalars = self._bagua_broadcast_scalars(scalars, src=0)
+        for key, p in scalars.items():
+            call_back_param[key](p)
+
+    def _bagua_broadcast_scalars(self, scalars, src):
+        # Serializes and broadcast scalars by converting them to "ByteTensor".
+        b = io.BytesIO()
+        cloudpickle.dump(scalars, b)
```
Comment (wangraying) [inline bagua/torch_api/distributed.py:176]: It seems pickle can do the same thing.

File: tests/torch_api/test_broadcast_state.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                            env,
+                            opt_class,
+                            opt_hyper_param,
+                        ),
+                    )
+                    p.start()
+                    processes.append(p)
+                for p in processes:
+                    p.join(timeout=60)
+
```
Comment (wangraying) [inline tests/torch_api/test_broadcast_state.py:143]: We can add a statement `p.exitcode == 0` here to check whether the process terminates normally.

File: tests/torch_api/test_broadcast_state.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    ]
+    optimizer_params = get_optimizer_param_values(bagua_optimizer)
+
+    # Put "model_params" in dimension 1, while "optimizer_params" in dimension 2.
+    bagua_params[rank][0].extend(model_params)
+    bagua_params[rank][1].extend(optimizer_params)
+
+
+class Test_Broadcast_Module(unittest.TestCase):
+    def test_broadcast_module(self):
```
Comment (wangraying) [inline tests/torch_api/test_broadcast_state.py:96]: add `@skip_if_cuda_not_available()` for gpu tests to skip running on cpu nodes.

File: bagua/torch_api/distributed.py (hunk 1 above)
Comment (Tengxu-Sun) [reply bagua/torch_api/distributed.py:117]: Thanks for your review. It seems some states will be initialized in the first 'optimizer.step' call like shown in https://github.com/pytorch/pytorch/blob/1b1f1e36b42fd9042ba7b7c37f4a717b6631edff/torch/optim/sgd.py#L105 and https://github.com/pytorch/pytorch/blob/1b1f1e36b42fd9042ba7b7c37f4a717b6631edff/torch/optim/adam.py#L86., so this should be necessary.

File: tests/torch_api/test_broadcast_state.py (hunk 3 above)
Comment (Tengxu-Sun) [reply tests/torch_api/test_broadcast_state.py:143]: ok, i will add it and confirm.

File: tests/torch_api/test_broadcast_state.py (hunk 4 above)
Comment (Tengxu-Sun) [reply tests/torch_api/test_broadcast_state.py:96]: ok

File: bagua/torch_api/distributed.py (hunk 2 above)
Comment (Tengxu-Sun) [reply bagua/torch_api/distributed.py:176]: According to the 2nd paragraph of https://github.com/cloudpipe/cloudpickle#overriding-pickles-serialization-mechanism-for-importable-constructs, cloudpickle seems more suitable for distributed execution environment. Maybe more safe？

**Pull request merged**